app/src/main/java/bg/zahov/app/ui/history/info/HistoryInfoViewModel.kt

```diff
@@ -1,146 +1,175 @@
+ * @property sets A list of sets performed for the exercise, represented as strings.
+ * @property oneRepMaxes A list of one-rep maxes for the exercise, represented as strings.
+ */
+data class ExerciseDetails(
+    val exerciseName: String,
+    val sets: List<String>,
+    val oneRepMaxes: List<String>,
+)
+
+/**
+ * Represents the detailed information of a workout history entry.
+ *
+ * @property id The unique identifier of the workout.
+ * @property workoutName The name of the workout.
+ * @property workoutDate The date when the workout was performed.
+ * @property duration The duration of the workout, represented as a string.
+ * @property volume The total volume lifted during the workout, represented as a string.
+ * @property prs The personal records achieved during the workout, represented as a string.
+ * @property exercisesInfo A list of detailed information for each exercise in the workout.
+ * @property isDeleted indicates whether the workout has been deleted
+ */
+data class HistoryInfoWorkout(
+    val id: String = "",
+    val workoutName: String = "",
+    val workoutDate: String = "",
+    val duration: String = "",
+    val volume: String = "",
+    val prs: String = "",
+    val exercisesInfo: List<ExerciseDetails> = listOf(),
+    val isDeleted: Boolean = false,
+)
+
+/**
+ * ViewModel for managing the state and actions related to workout history details.
+ *
+ * @constructor Creates a HistoryInfoViewModel with the required dependencies.
+ * @property workoutStateProvider Provides the current workout state.
+ * @property workoutProvider Handles operations related to workouts, such as retrieving and modifying them.
+ * @property toastManager Manages the display of toast notifications.
+ */
+class HistoryInfoViewModel(
+    private val workoutStateProvider: WorkoutStateManager = Inject.workoutState,
+    private val workoutProvider: WorkoutProvider = Inject.workoutProvider,
+    private val toastManager: ToastManager = ToastManager,
+) : ViewModel() {
+
+    private val _uiState = MutableStateFlow(HistoryInfoWorkout())
+    val uiState: StateFlow<HistoryInfoWorkout> = _uiState
+
     private var templates: MutableList<Workout> = mutableListOf()
+    private var pastWorkouts: MutableList<Workout> = mutableListOf()
     private var currentWorkoutState: WorkoutState? = null
-    val state: LiveData<State>
-        get() = _state
 
+    /**
+     * Initializes the ViewModel by collecting template workouts, the current workout state,
+     * and the selected workout from history.
+     */
     init {
         viewModelScope.launch {
             launch {
-                try {
-                    workoutProvider.getTemplateWorkouts().collect {
-                        templates = it.toMutableList()
-                    }
-                } catch (e: CriticalDataNullException) {
-                    serviceError.initiateCountdown()
+                workoutProvider.getTemplateWorkouts().collect {
```

Comment:
You can assign those directly to the flow fields instead of using the `init{}`. 
e.g. 
```
private val templates: MutableList<Workout> = workoutProvider.getTemplateWorkouts().launchIn(viewModelScope)
```

Reply:
That way you won't need to collect the flow in two places (the ViewModel and the view).

Reply:
The example I gave was just a pseudocode sample. To apply this to your use case you can replace 
```
private var templates: MutableList<Workout> = mutableListOf()
```
with:
```
private val templates: Flow<MutableList<Workout>> = workoutProvider.getTemplateWorkouts().map { it.toMutableList() }
```
or if you don't need it to be mutable with:
```
    private val templates: Flow<List<Workout>> = workoutProvider.getTemplateWorkouts()
```
and then you can remove this block:
```
     workoutProvider.getPastWorkouts().collect {
                    pastWorkouts = it.toMutableList()
                }
```

Reply:
I don't know if I am doing something wrong, however the only way I was able to implement this approach was like this:
`    private val templates: Flow<List<Workout>> by lazy {
        flow {
            workoutProvider.getTemplateWorkouts().collect {
                emit(it)
            }
        }
    }`
    
    This does not work because getTemplateWorkouts() is a suspend function and we are calling it outside of a suspend
    block when declaring templates
    `private val templates: Flow<MutableList<Workout>> = workoutProvider.getTemplateWorkouts().map { it.toMutableList() }`